Documentation/update-modules-v3-file-api.md

```diff
@@ -0,0 +1,422 @@
+
+### Marking augmented artifacts as supported
+
+If you wish to accept `augment` files you need to implement this calling
+interface:
+
+```bash
+./update-module SupportsAugmentedArtifacts
+```
+
+The module must respond by printing this exact string, followed by a newline:
+
+```
+YesThisModuleSupportsAugmentedArtifacts
+```
+
+and then return zero. All other output and return codes will be interpreted as
+not supporting augmented artifacts, and if any artifact arrives that has
+augmented headers or files, it will be rejected.
+
+
+### Determining which types of artifacts are supported
+
+When using augmented artifacts, sometimes the augmented artifact will have a
+different update `type` than the originally signed artifact, and the `type` is
+also the name of the update module. Since this makes the type untrusted, it is
+important that the update module listed in the augmented header knows how to
+handle an artifact with the `type` listed in the original header.
+
+Let's take an example, and call it `rootfs-delta-image`. This update module is
+used to package binary deltas from an original rootfs image. Therefore the
+update module knows how to handle artifacts whose original `type` was
+`rootfs-image`. To communicate this to Mender, the module is expected to answer
+the following call:
+
+```bash
+./update-module ListSupportedOriginalTypes
+```
+
+to which it should reply with a newline separated list of types it supports. In
+our example there is only one, so it responds by printing:
+
+```
+rootfs-image
+```
+
+This confirms that an artifact whose augmented `type` is `rootfs-delta-image` is
+an acceptable `type` for an original signed artifact whose `type` was
+`rootfs-image`.
+
+As examples of negative matches, for a bogus `type` value, the update module won't
+be found, and hence the update will fail. If listing a different update module,
+which exists, but is unrelated, that module will not list `rootfs-image` as
+being in its list of supported original types, and hence this update will also
+fail. Only when there is a match between the two will the `type` field be
+accepted.
+
+Note that in our `rootfs-delta-image` example, `rootfs-image` would also need to
+be prepared for this scenario, because its payload should be stored in the
+augmented section so that it can be removed in an augmented update in favor of
+the binary delta file.
+
+
+### Filtering header fields
+
+In general, augmented artifacts should not be allowed to override most headers
+from the original artifact, because this may be insecure. But some headers may
+need to be overrideable so that the update module can be provided with the
+context it needs to install the augmented artifact.
+
+To permit specific header fields to be overridden, Mender calls the update
+module like this:
+
+```bash
+./update-module PermittedAugmentedHeaders
+```
+
+The module is expected to return a JSON structure like this:
```

Comment:
JSON might be a bit tricky in shell/C, as we know.. maybe relates to my "manifest" note above.

Reply:
Wasn't the file api kind of an effort to be free from JSON?

Reply:
I think JSON is sort of unavoidable. The problem is that all the headers are JSON too, and some are free form, so the structure isn't necessarily known in advance. Therefore it's hard to represent in a flat model.

I think it's ok to use JSON here, given that this is advanced usage. Most update modules are not expected to deal with `PermittedAugmentedHeaders`, or even most of the headers that are stored in the file API. A package installer or app deployment module doesn't need any of this, for example. Besides, this is just printing, but if you return anything here, you will have to parse JSON as well, because that's what you will get in the artifact. Mender can't help you there, because it only knows its own headers.